src/election_works/voting_methods.clj

```
@@ -0,0 +1,81 @@
+(ns election-works.voting-methods
+  (:require [clojure.tools.logging :as log]
+            [election-works.utils :refer [maybe-assoc]]))
+
+(defn assoc-props
+  [{:keys [ballot-request-deadline
+           early-voting-start early-voting-end]}
+    voting-method]
+  (case (:type voting-method)
+    :by-mail
+    (maybe-assoc voting-method
+                 :ballot-request-deadline-received ballot-request-deadline)
+    :early-voting
+    (-> voting-method
+        (maybe-assoc :start early-voting-start)
+        (maybe-assoc :end early-voting-end))
+    voting-method))
+
+(defn merge-voting-methods
+  [defaults election-props]
+  (map (partial assoc-props election-props) defaults))
+
+(defn supported-voting-methods
+  "Keep only the supported voting-method types."
+  [supported-methods voting-methods]
+  (if supported-methods
+    (filter #(contains? supported-methods (:type %)) voting-methods)
+    voting-methods))
+
+(defn select-primary-voting-method
+  "If the primary-voting-method is specified, override the
+  settings currently in voting-methods.
+
+  If the current voting-methods doesn't include primary-voting-method, throw an exception.
+
+  When we set the primary-voting-method to be by-mail, we also set excuse-required
+  to false, and dissociate the acceptable-forms and ballot-request-deadline-received
+  properties (which might have been set based on state authority settings),
+  because a voter wouldn't need to request a ballot for an all-mail election.
+  "
+  [primary-voting-method voting-methods]
+  (case primary-voting-method
+    :by-mail
+    (if (some #(= :by-mail (:type %)) voting-methods)
+      (map #(if (= (:type %) :by-mail)
+              (assoc % :primary true)
+              (assoc % :primary false))
+           voting-methods)
+      (throw (Exception. (str ":by-mail is not a valid voting-method and cannot be made as the primary voting-method." ))))
+    voting-methods))
```

I'm wondering if it makes sense to have the logic be the same no matter the `primary-voting-method`, i.e. set that one as `:primary true` and all the other ones as `:primary false`, and throw if the method you're trying to set as primary doesn't exist.

I'm not sure we have a use-case right now for setting anything but `:by-mail` as the primary voting method, but if for instance, an election tried to set `:in-person` as the primary voting method, right now we're ignoring that.

---

yeah, I guess I just wasn't ready to generic set  the primary voting method to just anything yet.  Like, if we set `:by-mail` to be primary, we have to go back and make sure `excuse-required` is false, and get rid of the acceptable forms.  And I didn't know if we were ready to consider what it would mean, for example, if someone set `:early-voting` to be primary.

---

That motivation makes sense. I'm fine with this function handling primary-voting-methods differently, but it should throw if you give it a non-nil primary-voting-method that we don't know how to handle.

I believe the schema also considers something a `:primary` `:early-voting` method as invalid, so we could also handle it that (and if it doesn't, we should fix that). I'd say as much as possible the import code should handle whatever input you give it, but throw if it produces an invalid election -- and that could happen explicitly here, or later when we validate the whole election against the schema.

---

also fwiw unless the exception type you're throwing has some kind of specific meaning, I'd say it's a little more idiomatic to do:
```clj
(throw (ex-info "some message" {:some :relevant, :extra data}))
```